ev3dev2/motor.py

```diff
@@ -2099,6 +2111,169 @@ def follow_line(self,
 
         self.stop()
 
+    def calibrate_gyro(self):
+        """
+        Calibrates the gyro sensor.
+
+        NOTE: This takes 1sec to run
+        """
+        assert self.gyro, "GyroSensor must be defined"
+
+        for x in range(2):
+            self.gyro.mode = 'GYRO-RATE'
+            self.gyro.mode = 'GYRO-ANG'
+            time.sleep(0.5)
+
+    def follow_gyro_angle(self,
+            kp, ki, kd,
+            speed,
+            target_angle=0,
+            sleep_time=0.01,
+            follow_for=follow_for_forever,
+            **kwargs
+        ):
+        """
+        PID gyro angle follower
+
+        ``kp``, ``ki``, and ``kd`` are the PID constants.
+
+        ``speed`` is the desired speed of the midpoint of the robot
+
+        ``target_angle`` is the angle we want drive
+
+        ``sleep_time`` is how many seconds we sleep on each pass through
+            the loop.  This is to give the robot a chance to react
+            to the new motor settings. This should be something small such
+            as 0.01 (10ms).
+
+        ``follow_for`` is called to determine if we should keep following the
+            line or stop.  This function will be passed ``self`` (the current
+            ``MoveWithGyroTank`` object). Current supported options are:
+            - ``follow_for_forever``
+            - ``follow_for_ms``
+
+        ``**kwargs`` will be passed to the ``follow_for`` function
+
+        Example:
+
+        .. code:: python
+
+            from ev3dev2.motor import OUTPUT_A, OUTPUT_B, MoveTank, SpeedPercent, follow_for_ms
+            from ev3dev2.sensor.lego import GyroSensor
+            tank = MoveTank(OUTPUT_A, OUTPUT_B)
+            tank.gyro = GyroSensor()
+            try:
+                # Follow the line for 4500ms
+                tank.calibrate_gyro()
+                tank.follow_gyro_angle(
+                    kp=11.3, ki=0.05, kd=3.2,
+                    speed=SpeedPercent(30),
+                    follow_for=follow_for_ms,
+                    ms=4500
+                )
+            except Exception:
+                tank.stop()
+                raise
+        """
+        assert self.gyro, "GyroSensor must be defined"
+
+        integral = 0.0
+        last_error = 0.0
+        derivative = 0.0
+        speed_native_units = speed.to_native_units(self.left_motor)
+        MAX_SPEED = SpeedNativeUnits(self.max_speed)
+
+        while follow_for(self, **kwargs):
+            current_angle = self.gyro.angle
+            error = current_angle - target_angle
+            integral = integral + error
+            derivative = error - last_error
+            last_error = error
+            turn_native_units = (kp * error) + (ki * integral) + (kd * derivative)
+
+            left_speed = SpeedNativeUnits(speed_native_units - turn_native_units)
+            right_speed = SpeedNativeUnits(speed_native_units + turn_native_units)
+
+            if left_speed > MAX_SPEED:
```

Comment:
Looking at it... what's the motivation for failing if the speed is too high as opposed to just capping it at that max speed?

I think we did this for the line following too and not I'm not sure why.

Reply:
It's a good way for the user to know that they're potentially shooting themselves in the foot. Going over a max could result in the sensors and algorithm to not correct themselves within the feedback loop's quantum.